sdk/openai/Azure.AI.OpenAI/src/Custom/AzureOpenAIClientOptions.cs

```diff
@@ -39,5 +66,33 @@ public enum ServiceVersion
         V2024_06_01 = 9,
     }
 
+    internal class RetryWithDelaysPolicy : ClientRetryPolicy
+    {
+        protected override TimeSpan GetNextDelay(PipelineMessage message, int tryCount)
+        {
+            TimeSpan? TryGetTimeSpanFromHeader(string headerName, int millisecondsPerValue = 1, bool allowDateTimeOffset = false)
+            {
+                if (double.TryParse(
+                    message?.Response?.Headers?.TryGetValue(headerName, out string textValue) == true ? textValue : null,
+                    out double doubleValue) == true)
+                {
+                    return TimeSpan.FromMilliseconds(millisecondsPerValue * doubleValue);
+                }
+                else if (allowDateTimeOffset && DateTimeOffset.TryParse(headerName, out DateTimeOffset delayUntil))
+                {
+                    return delayUntil - DateTimeOffset.Now;
+                }
+                return null;
+            }
+
+            TimeSpan? delayFromHeader =
+                TryGetTimeSpanFromHeader("retry-after-ms")
+                ?? TryGetTimeSpanFromHeader("x-ms-retry-after-ms")
+                ?? TryGetTimeSpanFromHeader("Retry-After", millisecondsPerValue: 1000, allowDateTimeOffset: true);
+
+            return delayFromHeader ?? base.GetNextDelay(message, tryCount);
```

Comment:
Will this work with exponential backoff logic if the header delay is shorter than what the exponential backoff logic specifies?

Reply:
I believe we always want to honor the header value first; this is consistent with other OpenAI SDKs, at least, and I don't think we want overall client retry logic to second-guess what could and should be better-informed knowledge from the service.